### PR TITLE
Record segment generation of ongoing flush

### PR DESCRIPTION
**Pull request by pxsalehi**
The `LiveVersionMapArchive` records the segment generation that a set of doc IDs belong to and 
clears them once that generation is visible on the search shards. However, while a flush is ongoing 
(after committing `IndexWriter` and before updating the `lastCommittedSegmentInfos`) concurrent 
index/update operations that see a refresh and are moved to archive could incorrectly be considered 
as part of the propagated commit. This change records the ongoing commit generation slightly earlier 
to be able to safely handle this concurrent index/updates.

relates ES-5728

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 1]
```diff
@@ -145,8 +145,10 @@ public class InternalEngine extends Engine {
     private final LiveVersionMapArchive liveVersionMapArchive;
     // Records the last known generation during which LiveVersionMap was in unsafe mode. This indicates that only after this
     // generation it is safe to rely on the LiveVersionMap for a real-time get.
-    // TODO: can we move this entirely into the stateless plugin?
+    // TODO: move the following two to the stateless plugin
```
Comment (pxsalehi) [inline server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:148]: I think, extending the `LiveVersionMapArchive` interface to have a `preCommit` and `switchToSafeMap` hook, is probably one way of moving these out of here.

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 2]
```diff
@@ -145,8 +145,10 @@ public class InternalEngine extends Engine {
...
     private final AtomicLong lastUnsafeSegmentGenerationForGets = new AtomicLong(-1);
+    // Records the segment generation for the currently ongoing commit if any, or the last finished commit otherwise.
+    protected final AtomicLong preCommitRecordedSegmentGeneration = new AtomicLong(-1);
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:151]: Can we remove `Recorded` from the name? Does not seem to add anything significant information.

Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:151]: Can we make this private instead and add a protected accessor for the long value instead?

File: server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java [hunk 3]
```diff
@@ -673,6 +673,42 @@ protected void commitIndexWriter(IndexWriter writer, Translog translog) throws I
...
+            protected void commitIndexWriter(IndexWriter writer, Translog translog) throws IOException {
+                recordedGen.set(preCommitRecordedSegmentGeneration.get());
+                lastSegmentInfoGenUponCommit.set(getLastCommittedSegmentInfos().getGeneration());
+                super.commitIndexWriter(writer, translog);
+            }
+
+            @Override
+            public void flush(boolean force, boolean waitIfOngoing, ActionListener<FlushResult> listener) throws EngineException {
+                super.flush(force, waitIfOngoing, listener);
+                postFlushSegmentInfoGen.set(getLastCommittedSegmentInfos().getGeneration());
```
Comment (henningandersen) [inline server/src/test/java/org/elasticsearch/index/engine/InternalEngineTests.java:693]: Can we add an assertion here that the `preCommitRecordedSegmentGeneration` is still the same?

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 1 above)
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:148]: That makes sense to me.

**Review by henningandersen — APPROVED** [review]
LGTM.

**Comment by pxsalehi** [comment]
@elasticmachine run elasticsearch-ci/part-1

failure is https://github.com/elastic/elasticsearch/issues/95933, muting it in https://github.com/elastic/elasticsearch/pull/95937.

**Pull request merged**